charmcraft/main.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+
+import argparse
+import logging
+import operator
+import os
+import sys
+
+from charmcraft import logsetup
+from charmcraft.commands import version
+from charmcraft.cmdbase import CommandError
+
+logger = logging.getLogger(__name__)
+
+
+# Collect commands in different groups, for easier human consumption. Note that this is not
+# declared in each command because it's much easier to do this separation/grouping in one
+# central place and not distributed in several classes/files.
+COMMANDS_GROUPS = [
+    ('basic', "basics", [version.VersionCommand, version.CommandExampleError]),
+    ('advanced', "more complex stuff", [version.CommandExampleDebug]),
+]
+
+
+class CustomArgumentParser(argparse.ArgumentParser):
+    """ArgumentParser with grouped commands help."""
+
+    # Flag to indicate action groups that will have custom docs
+    special_group = 'MARKER'
+
+    def __init__(self, **kwargs):
+        self.__commands_groups = kwargs.pop('commands_groups', ())
+        super().__init__(**kwargs)
+
+    def format_help(self):
+        """Produce normal help but showing the commands grouped."""
+        main = False
+        for ag in self._action_groups:
+            if ag.title is self.special_group:
+                self._action_groups.remove(ag)
+                main = True
+                break
+        base = super().format_help()
+        if not main:
+            return base
+
+        # Get the size of the longest name so all helps are columned properly across the groups.
+        longest_name = 0
+        for _, _, cmd_classes in self.__commands_groups:
+            for cmd_class in cmd_classes:
+                longest_name = max(len(cmd_class.name), longest_name)
+
+        extra = ['', 'commands:', '']
+        for group, group_title, cmd_classes in self.__commands_groups:
+            extra.append("    {}:".format(group_title))
+            for cmd_class in sorted(cmd_classes, key=operator.attrgetter('name')):
```

Comment:
Right, I sorted them for easier human consumption. `snapcraft -h` does the same thing.

It may happen that in a future we may want to have a specific "handcrafted" order. In that case, we could change this, but that situation will arise only after having lots of commands (and Design involved), if ever.

So I propose to leave it like this, revisit in some months from now.

Reply:
:+1: